### ISSUE TITLE
KeyError: 'Binary' in Noetic import-package job

### ISSUE THREAD

**Issue by sloretz** [ros-infrastructure/reprepro-updater#73]
The ROS Noetic `Nrel_import-package` job appears to be failing with a python trace back http://build.ros.org/job/Nrel_import-package/2/console .
<details><summary>import debian package section</summary>

```
00:00:00.699 # BEGIN SECTION: import debian package
00:00:00.699 + export PYTHONPATH=/home/jenkins-agent/workspace/Nrel_import-package/reprepro-updater/src:
00:00:00.699 + python -u /home/jenkins-agent/workspace/Nrel_import-package/reprepro-updater/scripts/include_folder.py --folder /var/repos/ubuntu/building/queue/focal/Nsrc_uF__catkin__ubuntu_focal__source__1 --package ros-noetic-catkin --delete-folder --commit --invalidate
00:00:00.724 Traceback (most recent call last):
00:00:00.724   File "/home/jenkins-agent/workspace/Nrel_import-package/reprepro-updater/scripts/include_folder.py", line 84, in <module>
00:00:00.724     if options.package in c.content['Binary'].split()]
00:00:00.724   File "/usr/lib/python2.7/dist-packages/debian/deb822.py", line 245, in __getitem__
00:00:00.724     value = self.__dict[key]
00:00:00.724 KeyError: 'Binary'
00:00:00.732 Build step 'Execute shell' marked build as failure
00:00:00.738 [description-setter] Could not determine description.
00:00:00.744 Email was triggered for: Failure - Any
00:00:00.744 Sending email for trigger: Failure - Any
00:00:00.747 Sending email to: ros-buildfarm-noetic@googlegroups.com sloretz+buildfarm@openrobotics.org
00:00:00.813 Finished: FAILURE
```
</details>

That traceback appears to be coming from the `refactor` branch of this repository.

https://github.com/ros-infrastructure/reprepro-updater/blob/bc2962bcf0d7f0a9915720c09b89359ff3e38a87/scripts/include_folder.py#L83-L84

Which expects a `ChangesFile.content['Binary']` to always succeed.
https://github.com/ros-infrastructure/reprepro-updater/blob/bc2962bcf0d7f0a9915720c09b89359ff3e38a87/src/reprepro_updater/changes_parsing.py#L10

`ChangesFile.content` is type [debian.deb822.Changes](https://salsa.debian.org/python-debian-team/python-debian/blob/master/lib/debian/deb822.py#L1779).

https://github.com/ros-infrastructure/reprepro-updater/blob/bc2962bcf0d7f0a9915720c09b89359ff3e38a87/src/reprepro_updater/changes_parsing.py#L10

It's a little hard to follow, I'm not sure what keys `Changes` is expected to have. The [Debian changes files documentation](https://www.debian.org/doc/debian-policy/ch-controlfields.html#debian-changes-files-changes) does list `Binary` as a mandatory field, so it seems reasonable to expect that to exist on the Python representation. I don't know why it doesn't exist in this case.

**Comment by sloretz** [comment]
Related bugs: 

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=818618
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=922636
https://bugs.launchpad.net/launchpad/+bug/1813037

Looks like this is [an intentional change](https://git.dpkg.org/cgit/dpkg/dpkg.git/diff/?id=4a4619831).

```diff
+  * dpkg-genchanges: Only reference binary packages being uploaded, which
+    means that for a source-only upload, the Binary and Description fields
+    should be empty. Closes: #818618
```

**Comment by sloretz** [comment]
Since the code in this repo is using `Binary` field to [filter unrelated `.changes` files](https://github.com/ros-infrastructure/reprepro-updater/blob/bc2962bcf0d7f0a9915720c09b89359ff3e38a87/scripts/include_folder.py#L83-L84), can we use [the Source field](https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-source) instead? The code is filtering for [only one package](https://github.com/ros-infrastructure/reprepro-updater/blob/bc2962bcf0d7f0a9915720c09b89359ff3e38a87/scripts/include_folder.py#L57), and at least in catkin's case the `Source` field seems to match the value passed as `--package` which is `ros-noetic-catkin`.

```
python -u /home/jenkins-agent/workspace/Nrel_import-package/reprepro-updater/scripts/include_folder.py --folder /var/repos/ubuntu/building/queue/focal/Nsrc_uF__catkin__ubuntu_focal__source__1 --package ros-noetic-catkin --delete-folder --commit --invalidate
```

```
Format: 1.8
Date: Tue, 10 Dec 2019 08:00:00 -0000
Source: ros-noetic-catkin
Architecture: source
Version: 0.7.20-0focal
Distribution: focal
Urgency: high
Maintainer: Dirk Thomas <dthomas@osrfoundation.org>
Changed-By: Dirk Thomas <dthomas@osrfoundation.org>
Changes:
 ros-noetic-catkin (0.7.20-0focal) focal; urgency=high
 .
   * fix checking dependency conditions (#1041 <https://github.com/ros/catkin/issues/1041>)
   * fix gtest-not-found bug, regression from 0.7.19 (#1040 <https://github.com/ros/catkin/issues/1040>)
   * support build_type tag with condition attribute (#1038 <https://github.com/ros/catkin/issues/1038>)
   * use ${prefix} variable in generated pkg-config (#1037 <https://github.com/ros/catkin/issues/1037>)
   * check for INTERFACE library type in add_library function (#1034 <https://github.com/ros/catkin/issues/1034>)
Checksums-Sha1:
 b7e65542bc650121c6472691a73e602a71306e56 1016 ros-noetic-catkin_0.7.20-0focal.dsc
 9b8c309ada57bce50a9f9303da108c99f10b3baa 214757 ros-noetic-catkin_0.7.20.orig.tar.gz
 ab8e610da823ce4f64c9d02e6bd11d2f84289e8c 11068 ros-noetic-catkin_0.7.20-0focal.debian.tar.xz
 152c05d4ecd3c19eed53e56aa67fa44e1c05ef3a 4754 ros-noetic-catkin_0.7.20-0focal_source.buildinfo
Checksums-Sha256:
 d42597e6486abd559268a1818055c3fe708942efee47d0a6e392527151882d41 1016 ros-noetic-catkin_0.7.20-0focal.dsc
 d328e25ade9aeb031b2ba05951f6860842e687567c7185262aca9865da1e85c3 214757 ros-noetic-catkin_0.7.20.orig.tar.gz
 16f2e138c500119e8ee4a5ec658100938f4b6d72a11b645dd1e9952ebcd2b89c 11068 ros-noetic-catkin_0.7.20-0focal.debian.tar.xz
 f54e0c6273a9661afa8edea2c0e0dccea155436ac686090d89e29cd8583c519f 4754 ros-noetic-catkin_0.7.20-0focal_source.buildinfo
Files:
 f85f06d8bb8eb936ed1872fcbdff3b49 1016 misc optional ros-noetic-catkin_0.7.20-0focal.dsc
 540d256f3acf243dd2723e9e168727bb 214757 misc optional ros-noetic-catkin_0.7.20.orig.tar.gz
 3e7436b23fcf6ca97cb46ad5904963ba 11068 misc optional ros-noetic-catkin_0.7.20-0focal.debian.tar.xz
 2f36e15062f20454daa08565e17c6c9b 4754 misc optional ros-noetic-catkin_0.7.20-0focal_source.buildinfo
```

**Comment by tfoote** [comment]
With #74 in the problem is just one step down the line at reprepro rejecting it due to the missing field.

It's "fixed" in the "latest" version https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=920377

```
16:09:54 # BEGIN SECTION: import debian package
16:09:54 + export PYTHONPATH=/home/jenkins-agent/workspace/Nrel_import-package/reprepro-updater/src:
16:09:54 + python -u /home/jenkins-agent/workspace/Nrel_import-package/reprepro-updater/scripts/include_folder.py --folder /var/repos/ubuntu/building/queue/buster/Nsrc_dB__catkin__debian_buster__source__3 --package ros-noetic-catkin --delete-folder --commit --invalidate
16:09:54 running command ['reprepro', '-b', '/var/repos/ubuntu/building', '-T', 'dsc', '-V', 'removefilter', u'buster', 'Package (== ros-noetic-catkin )']
16:09:55 running ['reprepro', '-v', '-b', '/var/repos/ubuntu/building', 'deleteunreferenced']
16:09:55 running command ['reprepro', '-v', '-b', '/var/repos/ubuntu/building', 'deleteunreferenced']
16:09:57 Importing package: ros-noetic-catkin:0.7.20-0buster:buster:source
16:09:57 running command ['reprepro', '-v', '-b', '/var/repos/ubuntu/building', '-S', 'misc', 'include', u'buster', '/var/repos/ubuntu/building/queue/buster/Nsrc_dB__catkin__debian_buster__source__3/ros-noetic-catkin_0.7.20-0buster_source.changes']
16:09:58 In '/var/repos/ubuntu/building/queue/buster/Nsrc_dB__catkin__debian_buster__source__3/ros-noetic-catkin_0.7.20-0buster_source.changes': Missing 'Binary' field
16:09:58 There have been errors!
16:09:58 Execution of [['reprepro', '-v', '-b', '/var/repos/ubuntu/building', '-S', 'misc', 'include', u'buster', '/var/repos/ubuntu/building/queue/buster/Nsrc_dB__catkin__debian_buster__source__3/ros-noetic-catkin_0.7.20-0buster_source.changes']] Failed: Command '['reprepro', '-v', '-b', '/var/repos/ubuntu/building', '-S', 'misc', 'include', u'buster', '/var/repos/ubuntu/building/queue/buster/Nsrc_dB__catkin__debian_buster__source__3/ros-noetic-catkin_0.7.20-0buster_source.changes']' returned non-zero exit status 255
16:09:58 Build step 'Execute shell' marked build as failure
16:09:58 [description-setter] Could not determine description.
16:09:58 Email was triggered for: Failure - Any
```

**Comment by tfoote** [comment]
The issue on this repo is resolved by #74 so i'm going to close this. I've opened https://github.com/ros-infrastructure/buildfarm_deployment/issues/223 to track the next iteration of the problem.
